src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java

```diff
@@ -7,40 +7,45 @@
 import com.onarandombox.MultiverseCore.event.MVPlayerTouchedPortalEvent;
 import com.onarandombox.MultiverseCore.utils.PermissionTools;
 import com.onarandombox.MultiverseNetherPortals.MultiverseNetherPortals;
-import com.onarandombox.MultiverseNetherPortals.enums.PortalType;
+import com.onarandombox.MultiverseNetherPortals.runnables.PlayerTouchingPortalTask;
 import com.onarandombox.MultiverseNetherPortals.utils.MVLinkChecker;
 import com.onarandombox.MultiverseNetherPortals.utils.MVNameChecker;
 import org.bukkit.Location;
 import org.bukkit.Material;
+import org.bukkit.PortalType;
 import org.bukkit.block.Block;
 import org.bukkit.block.BlockFace;
 import org.bukkit.entity.Player;
 import org.bukkit.event.EventHandler;
 import org.bukkit.event.EventPriority;
 import org.bukkit.event.Listener;
 import org.bukkit.event.entity.EntityPortalEnterEvent;
+import org.bukkit.event.entity.EntityPortalExitEvent;
+import org.bukkit.scheduler.BukkitTask;
 import org.bukkit.util.Vector;
 
 import java.util.Date;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Map;
+import java.util.Set;
 import java.util.logging.Level;
 
 public class MVNPEntityListener implements Listener {
 
-    private MultiverseNetherPortals plugin;
-    private MVNameChecker nameChecker;
-    private MVLinkChecker linkChecker;
-    private MVWorldManager worldManager;
-    private PermissionTools pt;
-    private int cooldown = 250;
-    private MultiverseMessaging messaging;
-    private Map<String, Date> playerErrors;
-    private Map<String, Location> eventRecord;
-    private LocationManipulation locationManipulation;
-    // This hash map will track players most recent portal touch.
-    // we can use this cache to avoid a TON of unrequired calls to the
-    // On entity portal touch calculations.
+    private final MultiverseNetherPortals plugin;
+    private final MVNameChecker nameChecker;
+    private final MVLinkChecker linkChecker;
+    private final MVWorldManager worldManager;
+    private final PermissionTools pt;
+    private final int cooldown = 250;
+    private final MultiverseMessaging messaging;
+    private final Map<String, Date> playerErrors;
+    private final LocationManipulation locationManipulation;
+    private final Map<PortalType, Set<Player>> eventRecord;
```

Comment:
We definitely do not want to use `Player` objects here. Storing player objects in a persistent variable can cause a lot of issues (e.g. ghost players). It would probably be better to use UUID here.

Reply:
Good point.